// qawolf/build/build-code/buildStepLines.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildStepLines = exports.buildExpressionLine = exports.buildValue = exports.escapeSelector = exports.getStepPageVariableName = void 0;
const util_1 = require("util");
/**
 * @summary Given a step, returns the correct page variable for it,
 *   such as `page` for the main page or `page2` for the second page.
 */
exports.getStepPageVariableName = (step) => {
    const { page } = step.event;
    return `page${page === 0 ? '' : page + 1}`;
};
exports.escapeSelector = (selector) => {
    if (!selector.includes(`"`))
        return `"${selector}"`;
    if (!selector.includes(`'`))
        return `'${selector}'`;
    return '`' + selector.replace(/`/g, '\\`') + '`';
};
exports.buildValue = ({ action, value }) => {
    if (action === 'scroll') {
        const scrollValue = value;
        return `{ x: ${scrollValue.x}, y: ${scrollValue.y} }`;
    }
    if (util_1.isUndefined(value))
        return '';
    return JSON.stringify(value);
};
exports.buildExpressionLine = (step, frameVariable) => {
    const { action, event } = step;
    const args = [exports.escapeSelector(event.selector)];
    const value = exports.buildValue(step);
    if (value)
        args.push(value);
    const browsingContext = frameVariable || exports.getStepPageVariableName(step);
    let methodOpen = `${browsingContext}.${action}(`;
    if (action === 'scroll') {
        methodOpen = `qawolf.scroll(${browsingContext}, `;
    }
    const expression = `await ${methodOpen}${args.join(', ')});`;
    return expression;
};
exports.buildStepLines = (step, buildContext = {
    initializedFrames: new Map(),
    initializedPages: new Set(),
}) => {
    const lines = [];
    const { frameSelector, page } = step.event;
    const { initializedFrames, initializedPages } = buildContext;
    const pageVariableName = exports.getStepPageVariableName(step);
    if (page > 0 && !initializedPages.has(page)) {
        lines.push(`const ${pageVariableName} = await qawolf.waitForPage(page.context(), ${page});`);
        initializedPages.add(page);
    }
    let frameVariableName;
    if (frameSelector) {
        frameVariableName = initializedFrames.get(frameSelector);
        if (!frameVariableName) {
            frameVariableName = `frame${initializedFrames.size ? initializedFrames.size + 1 : ''}`;
            lines.push(`const ${frameVariableName} = await (await ${pageVariableName}.$(${exports.escapeSelector(frameSelector)})).contentFrame();`);
            initializedFrames.set(frameSelector, frameVariableName);
        }
    }
    lines.push(exports.buildExpressionLine(step, frameVariableName));
    return lines;
};
