// qawolf/build/qawolf.js
"use strict";
// qawolf API
Object.defineProperty(exports, "__esModule", { value: true });
var create_1 = require("./create-code/create");
Object.defineProperty(exports, "create", { enumerable: true, get: function () { return create_1.create; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "launch", { enumerable: true, get: function () { return utils_1.launch; } });
Object.defineProperty(exports, "register", { enumerable: true, get: function () { return utils_1.register; } });
Object.defineProperty(exports, "repl", { enumerable: true, get: function () { return utils_1.repl; } });
Object.defineProperty(exports, "saveState", { enumerable: true, get: function () { return utils_1.saveState; } });
Object.defineProperty(exports, "scroll", { enumerable: true, get: function () { return utils_1.scroll; } });
Object.defineProperty(exports, "setState", { enumerable: true, get: function () { return utils_1.setState; } });
Object.defineProperty(exports, "stopVideos", { enumerable: true, get: function () { return utils_1.stopVideos; } });
Object.defineProperty(exports, "waitForPage", { enumerable: true, get: function () { return utils_1.waitForPage; } });
