// create-qawolf/build/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.writeConfig = exports.detectYarn = exports.detectTypeScript = void 0;
const tslib_1 = require("tslib");
const fs_extra_1 = require("fs-extra");
const glob_1 = tslib_1.__importDefault(require("glob"));
const path_1 = require("path");
const cli_1 = require("./cli");
exports.detectTypeScript = () => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    return new Promise((resolve) => {
        glob_1.default('tsconfig*.json', (error, files) => {
            if (error)
                resolve(false);
            else
                resolve(files.length > 0);
        });
    });
});
exports.detectYarn = () => (process.env.npm_execpath || '').includes('yarn');
exports.writeConfig = ({ rootDir, useTypeScript, }) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const jestConfig = useTypeScript
        ?
            'node_modules/qawolf/ts-jest.config.json'
        : 'node_modules/qawolf/js-jest.config.json';
    const configFile = `module.exports = {
  config: "${jestConfig}",
  rootDir: "${rootDir}",
  testTimeout: 60000,
  useTypeScript: ${useTypeScript}
}\n`;
    const configPath = path_1.resolve('qawolf.config.js');
    if (!(yield cli_1.promptOverwrite(configPath)))
        return;
    yield fs_extra_1.writeFile(configPath, configFile, 'utf8');
});
