// qawolf/build/cli/createCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildCreateCommand = exports.runCreate = void 0;
const commander_1 = require("commander");
const config_1 = require("../config");
const parseUrl_1 = require("./parseUrl");
const buildEditOptions_1 = require("../run/buildEditOptions");
const runTests_1 = require("../run/runTests");
const saveTemplate_1 = require("./saveTemplate");
exports.runCreate = async (options) => {
    const config = config_1.loadConfig();
    const testPath = await saveTemplate_1.saveTemplate({
        device: options.device,
        name: options.name,
        rootDir: config.rootDir,
        statePath: options.statePath,
        templateFn: config.createTemplate,
        url: options.url,
        useTypeScript: config.useTypeScript,
    });
    if (!testPath) {
        // the user decided to not overwrite
        return;
    }
    runTests_1.runTests(buildEditOptions_1.buildEditOptions({
        args: options.args,
        config,
        env: {
            // discard should delete the test and selectors
            QAW_CREATE: 'true',
        },
        testPath,
    }));
};
exports.buildCreateCommand = () => {
    const command = new commander_1.Command('create')
        .storeOptionsAsProperties(false)
        .arguments('[url] [name]')
        .option('-d, --device <device>', 'emulate using a playwright.device')
        .option('--name <name>', 'name')
        .option('--statePath <statePath>', 'path where state data (cookies, localStorage, sessionStorage) is saved')
        .option('--url <url>', 'url')
        .description('✨ create a test from browser actions')
        .action(async () => {
        const opts = command.opts();
        const [urlArgument, nameArgument] = command.args;
        const urlString = opts.url || urlArgument || 'http://example.org';
        const url = parseUrl_1.parseUrl(urlString);
        let name = opts.name || nameArgument;
        if (!name) {
            name = (url.hostname || '').replace(/\..*/g, '');
        }
        await exports.runCreate({
            device: opts.device,
            name,
            statePath: opts.statePath,
            url: url.href,
        });
    });
    return command;
};
